[FluidDynamicsApplication] Local mass conservation process

swenczowski opened the pull request:
In the following, a process is suggested that provides a compensation for mass losses in two-fluid simulations. The aim was to localize the mass corrections by solving an additional convection problem:

Motivation:

- The global mass conservation process was found to work in a stable and reliable fashion
- The global mass correction distorts simulation results when (a) strongly local losses appear, (b) different water bodies exist within the domain, and (c) shallow water flow (e.g. just 1,5 elements filled) is present.

Approach:

- As before, the theoretical value of the global water volume is tracked by temporal integration of inflow and outflow
- The difference between the existing water volume and the expected value is computed
- The surface of the water domain is considered stationary and the (theoretical) flows through the surface are computed. Out of this observation, one can derive a time span that the convection must continue to achieve the desired water volume:

`dt = V_err / Q_surf`

- The current velocity field is used to convect the level-set field a bit further within the time step `dt.` By taking the minimum (maximum) between the actual distance field and the convected distance field, the volume `V_err` can be added (subtracted)
- Optionally, this procedure can be combined with the global conservation which is based an a global shift of the distance field.

Aims:

- computationally cheap procedure
- no further historical variables
- compatibility with global correction
- MPI implemented and OpenMP directives are inserted

Test case:

- Dam break scenario in deliberately coarse gridded domain with separated water basin. The simulation covers 5 s of physical time. The initial configuration is shown below whereas the further plots show the position of the water surface at the end of the simulation.
![initial](https://user-images.githubusercontent.com/45034120/52360247-de264d80-2a3b-11e9-8cb8-e0ffcc98ea1f.png)

- Extreme mass losses are observed if no correction is applied:
![0_withouta](https://user-images.githubusercontent.com/45034120/52360313-fdbd7600-2a3b-11e9-97f8-5da5d9e4a95f.png)

- If the global conservation is added, the mass losses in the "dynamic part" of the domain are reduced. Still, the missing volume in this part leads to a rise of the the water level in the separated basin. Mass in conserved globally.
![1_global](https://user-images.githubusercontent.com/45034120/52360463-660c5780-2a3c-11e9-836e-9631ca130d9c.png)

- With the local conservation, both separated water bodies show exact mass conservation. Splashes in the dam break scenario appear to be better resolved.
![2_local](https://user-images.githubusercontent.com/45034120/52360472-6c9acf00-2a3c-11e9-94fa-d9c53e505808.png)

- Running both corrections shows only minor differences to the case where only local conservation is applied. A smoothing effect of the global conservation can be observed.
![3_both](https://user-images.githubusercontent.com/45034120/52360519-8f2ce800-2a3c-11e9-8d1c-ecb5c383110b.png)

swenczi committed:
Modification of level_set_convection

swenczi committed:
Modification in Mass conservation process

swenczi committed:
Forward convection and "mass generation" generation working

swenczi committed:
Working version 2D

swenczi committed:
Working 2D serial version with complete interface

swenczi committed:
Adding documentation

swenczi committed:
Introducing MPI communication and OMP

swenczi committed:
Convection process with Trilinos process for convection (new class)

swenczi committed:
Adding MPI process to Python script

swenczi committed:
Using references in ApplyLocalCorrection()

swenczi committed:
Using references in OrthogonalFlowIntoAir()

RiccardoRossi commented on kratos/processes/levelset_convection_process.h:
```diff
@@ -153,19 +153,19 @@ class LevelSetConvectionProcess
+    ~LevelSetConvectionProcess() override
     {
-        mrBaseModelPart.GetModel().DeleteModelPart("DistanceConvectionPart");
+        // mrBaseModelPart.GetModel().DeleteModelPart("DistanceConvectionPart");
```
this file is used a lot,  and is "in production". Could u comment on what is the change you propose? i am also cc-ing @pablobecker since he is one of the users

RiccardoRossi commented on kratos/processes/levelset_convection_process.h:
```diff
@@ -390,19 +498,84 @@ class LevelSetConvectionProcess
+        KRATOS_CATCH("")
+    }
+
+
```
not really descriptive name ... what does this   function do?

swenczi committed:
Considering naming convention

swenczi committed:
Updating the default parameters and including check for empty model part

swenczi committed:
Merge remote-tracking branch 'origin/master' into feature-fldyn-local-mass-conservation

swenczi committed:
Change varriable to AUX_DISTANCE

swenczi committed:
Separation of MPI version (overrides the convection process setting)

swenczi committed:
Reverting all changes to the "old" level_set_convection_process

swenczi committed:
Cleaning, convention and including MPI basis in serial version

swenczi committed:
Inheriting everything but the constructor in MPI version

swenczi committed:
Documentation and removal of control output variable

swenczi committed:
Chaning defaults in C++ process

swenczowski commented:
As discussed with @rubenzorrilla, a number of changes has been made. The most prominent among these are:

- separation of the serial and the MPI version in the Python implementation.

- separation of the applied convection process (now named `level_set_forward_convection_process` and its MPI equivalent) from the original implementation of the `level_set_convection_process`. The reason was that nearly all functions had to be overridden. Further, maintaining the `level_set_convection_process` at its original implementation also addresses the remarks of @RiccardoRossi reading

> this file is used a lot, and is "in production". Could u comment on what is the change you propose? i am also cc-ing @pablobecker since he is one of the users

> not really descriptive name ... what does this function do?

- a documentation paragraph was added specifying the meaning of the `level_set_forward_convection_process`. The naming may not be optimal but also the search for convincing alternatives remained without success so far.

rubenzorrilla reviewed (commented):
@swenczowski I'll approve it as soon as the conflict is solved.

rubenzorrilla committed:
Merge branch 'master' into feature-fldyn-local-mass-conservation

philbucher commented:
ping is this still active?

rubenzorrilla commented:
Yes. I think I solved the conflict, but someone has to approve it.

philbucher commented:
ping @jcotela

ddiezrod reviewed (dismissed):
@rubenzorrilla This looks fine to me. I approve it but you can wait for @jcotela review if you prefer so

jcotela commented:
Go ahead and merge if you think it is ready. :+1:

Commit:
merging master

ddiezrod commented:
I forgot about this, can you re-approve it @rubenzorrilla @jcotela ?

Commit:
making intrusive

Commit:
move file to correct folder

philbucher reviewed (changes requested):
Can you please add some tests?
This PR adds file to important parts of the code

Or are there tests already?
thx

ddiezrod commented:
I am not sure, I think that there was one python test but for sure coverage can be improved. I will take a look

Commit:
Merge branch 'master' into feature-fldyn-local-mass-conservation

Commit:
vol 2

ddiezrod commented:
I did some important changes with the main goal of getting rid of the level set forward process. I also added some tests which can be run both in omp and mpi. @philbucher @rubenzorrilla

philbucher commented on kratos/processes/levelset_forward_convection_process.h:
```diff
@@ -0,0 +1,581 @@
+public:
+
+    KRATOS_DEFINE_LOCAL_FLAG(PERFORM_STEP1);
+    KRATOS_DEFINE_LOCAL_FLAG(DO_EXPENSIVE_CHECKS);
```
Why not implementing this in `Check`?

philbucher commented on kratos/processes/levelset_forward_convection_process.h:
```diff
@@ -0,0 +1,581 @@
+{
+public:
+
+    KRATOS_DEFINE_LOCAL_FLAG(PERFORM_STEP1);
```
maybe a more descriptive name?

philbucher commented on kratos/processes/levelset_forward_convection_process.h:
```diff
@@ -0,0 +1,581 @@
+ *
+*/
+template< unsigned int TDim, class TSparseSpace, class TDenseSpace, class TLinearSolver >
+class LevelSetForwardConvectionProcess : public Process
```
this has no check in the Core

I see that there is one in Fluid, but cannot this be used standalone?

philbucher reviewed (commented):
Minor comments
Also build fails

ddiezrod commented:
@philbucher What I tried in the last commits it's precisely to get completely rid of this forward level set process as I think the usual levelset convection process can be used here. (This new process was created in this MR some months ago by @swenczowski ) I have not removed it yet, until we agree if there is need for it or not.

ddiezrod commented on applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py:
```diff
@@ -54,12 +81,98 @@ def ExecuteBeforeSolutionLoop(self):
+                KratosFluid.AUX_DISTANCE,
+                self._fluid_model_part.GetCommunicator().LocalMesh().Nodes)
+            self.forward_convection_process.Execute()
+            self._fluid_model_part.ProcessInfo[KratosMultiphysics.DELTA_TIME] = prev_dt
```
These lines 108-115 are the main difference with respect to the first version. Distance before convection is stored in AUX_DISTANCE and then compared with the convected DISTANCE. Also, delta time is changed here to solve the convection step and then, the previous value is recovered.

philbucher commented:
> I have not removed it yet, until we agree if there is need for it or not.

then is the first step or am I missing sth?
I mean, why adding sth if we are not even sure if it is needed?

I will release by block but this confuses me...

philbucher commented on applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.h:
```diff
@@ -67,6 +67,10 @@ class KRATOS_API(FLUID_DYNAMICS_APPLICATION) MassConservationCheckProcess : publ
 
+    // typedef UblasSpace<double, CompressedMatrix, Vector> SparseSpaceType;
+    // typedef UblasSpace<double, Matrix, Vector> LocalSpaceType;
+    // typedef LinearSolver<SparseSpaceType, LocalSpaceType > LinearSolverType;
```
please remove commented code

philbucher commented on applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp:
```diff
@@ -32,129 +32,199 @@ namespace Kratos
+
+    // a small step is set to avoid numerical problems
+    double time_step_for_convection = 1.0e-7;
+    const auto& comm = mrModelPart.GetCommunicator();
```
```suggestion
    const auto& r_comm = mrModelPart.GetCommunicator();
```

philbucher commented on applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp:
```diff
@@ -32,129 +32,199 @@ namespace Kratos
+    double time_step_for_convection = 1.0e-7;
+    const auto& comm = mrModelPart.GetCommunicator();
+
+    if ( mWaterVolumeError > 0.0 ){
```
if this is not synced in MPI it turns into a deadlock

philbucher commented on applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp:
```diff
@@ -32,129 +32,199 @@ namespace Kratos
+                time_step_for_convection = mWaterVolumeError / water_outflow_over_boundary;
+            }
+        } else {
+            KRATOS_DEBUG_ERROR << "Communication failed in MassConservationCheckProcess::ComputeDtForConvection()";
```
```suggestion
            KRATOS_ERROR << "Communication failed in MassConservationCheckProcess::ComputeDtForConvection()";
```

philbucher commented on applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp:
```diff
@@ -32,129 +32,199 @@ namespace Kratos
+                time_step_for_convection = - mWaterVolumeError / water_inflow_over_boundary;
+            }
+        } else {
+            KRATOS_DEBUG_ERROR << "Communication failed in MassConservationCheckProcess::ComputeDtForConvection()";
```
```suggestion
            KRATOS_ERROR << "Communication failed in MassConservationCheckProcess::ComputeDtForConvection()";
```

philbucher commented on applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp:
```diff
@@ -32,129 +32,199 @@ namespace Kratos
+        // case: water volume was lost by mistake
+        mAddWater = true;
+        double water_outflow_over_boundary = OrthogonalFlowIntoAir( 1.0 );
+        comm.GetDataCommunicator().Barrier();
```
are you sure you need the barriers?

philbucher commented on applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp:
```diff
@@ -32,129 +32,199 @@ namespace Kratos
+        // case: water volume was gained by mistake
+        mAddWater = false;
+        double water_inflow_over_boundary = OrthogonalFlowIntoAir( -1.0 );
+        comm.GetDataCommunicator().Barrier();
```
same

philbucher commented on applications/FluidDynamicsApplication/tests/test_mass_conservation_check_process.py:
```diff
@@ -0,0 +1,229 @@
+    import KratosMultiphysics.TrilinosApplication as KratosTrilinos
+    import KratosMultiphysics.mpi.distributed_import_model_part_utility as distributed_import_model_part_utility
+except:
+    print("Running tests without MPI")
```
please improve those imports
Only import what you actually use (I am pretty sure you don't directly use metis)

Also use [this](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MappingApplication/tests/mapper_test_case.py#L49) for checking in a test if you are running MPI

philbucher commented on applications/FluidDynamicsApplication/tests/test_mass_conservation_check_process.py:
```diff
@@ -0,0 +1,229 @@
+            "model_import_settings" : {
+                "input_type" : "mdpa",
+                "input_filename" :\""""+ name +"""\",
+                "partition_in_memory" : """ + use_memory + """
```
this doesn't work in Win unfortunately

ddiezrod replied:
it does, I actually program in windows most of the time.

philbucher replied:
strange
I use it like this [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MappingApplication/tests/mapper_test_case.py#L147-L156) and this doesn't work in Win

do you know why?

ddiezrod replied:
This also works for me in windows. In fact, we have been using this for months in @KratosMultiphysics/altair and it has worked for everybody...

ddiezrod commented:
@philbucher I think you are right. I did not want to completely change the MR because it was not originally created by me, but it is confusing for someone that did not follow the whole process. I think I will create a new MR with the complete proposal.
